src/connector/src/sink/mqtt.rs

```diff
@@ -0,0 +1,262 @@
+            .map_err(SinkError::Mqtt)?;
+
+        Ok(())
+    }
+
+    async fn new_log_sinker(&self, writer_param: SinkWriterParam) -> Result<Self::LogSinker> {
+        Ok(MqttSinkWriter::new(
+            self.config.clone(),
+            self.schema.clone(),
+            self.pk_indices.clone(),
+            &self.format_desc,
+            self.db_name.clone(),
+            self.sink_from_name.clone(),
+            writer_param.executor_id,
+        )
+        .await?
+        .into_log_sinker(usize::MAX))
+    }
+}
+
+impl MqttSinkWriter {
+    pub async fn new(
+        config: MqttConfig,
+        schema: Schema,
+        pk_indices: Vec<usize>,
+        format_desc: &SinkFormatDesc,
+        db_name: String,
+        sink_from_name: String,
+        id: u64,
+    ) -> Result<Self> {
+        let formatter = SinkFormatterImpl::new(
+            format_desc,
+            schema.clone(),
+            pk_indices.clone(),
+            db_name,
+            sink_from_name,
+            &config.common.topic,
+        )
+        .await?;
+
+        let qos = config.common.qos();
+
+        let (client, mut eventloop) = config
+            .common
+            .build_client(0, id)
+            .map_err(|e| SinkError::Mqtt(anyhow!(e)))?;
+
+        let stopped = Arc::new(AtomicBool::new(false));
+        let stopped_clone = stopped.clone();
+        tokio::spawn(async move {
+            while !stopped_clone.load(std::sync::atomic::Ordering::Relaxed) {
+                match eventloop.poll().await {
+                    Ok(_) => (),
+                    Err(err) => match err {
+                        ConnectionError::Timeout(_) => (),
+                        ConnectionError::MqttState(rumqttc::v5::StateError::Io(err))
+                        | ConnectionError::Io(err)
+                            if err.kind() == std::io::ErrorKind::ConnectionAborted
+                                || err.kind() == std::io::ErrorKind::ConnectionReset =>
+                        {
+                            continue;
+                        }
+                        err => {
+                            tracing::error!("Failed to poll mqtt eventloop: {}", err.as_report());
+                            std::thread::sleep(std::time::Duration::from_secs(1));
+                        }
+                    },
+                }
+            }
+        });
+
+        let payload_writer = MqttSinkPayloadWriter {
+            client,
+            config: config.clone(),
+            qos,
+            retain: config.retain,
+        };
+
+        Ok::<_, SinkError>(Self {
+            config: config.clone(),
+            payload_writer,
+            schema: schema.clone(),
+            stopped,
+            formatter,
+        })
+    }
+}
+
+impl AsyncTruncateSinkWriter for MqttSinkWriter {
+    async fn write_chunk<'a>(
+        &'a mut self,
+        chunk: StreamChunk,
+        _add_future: DeliveryFutureManagerAddFuture<'a, Self::DeliveryFuture>,
```

Comment:
Here you didn't add any delivery future, which indicates that the chunk will have been delivered after the method returns. However, from the the doc of [AsyncClient::publish](https://docs.rs/rumqttc/latest/rumqttc/struct.AsyncClient.html#method.publish), after the method returns, IIUC, the message is only added to an event loop buffer, and has not been received by the external queue. If so, this may cause data loss if failure happens. 

From the doc, it seems that you should [poll](https://docs.rs/rumqttc/latest/rumqttc/struct.EventLoop.html#method.poll) the event loop to get notified on the successful delivery. If so, a better implementation might be, to implement a `MqttLogSinker` specially for mqtt sink, in which you use `select!` to poll both the upstream input from log_reader and delivery notification from mqtt client event loop. On delivery, you then truncate the log reader.

Reply:
@wenym1 This is not possible to implement because there's currently no way to map the published message with the `pkid` returned by the event loop, there's an issue already reported in [rumqtt](https://github.com/bytebeamio/rumqtt/issues/349) and an [RFC](https://github.com/bytebeamio/rumqtt/issues/805) to add the future for when the message is actually published